File: ARRAY/Multiplication/Multiplication.js
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Function to read matrix dimensions
function getMatrixDimensions(question) {
    return new Promise(resolve => {
        rl.question(question, input => {
            let values = input.trim().split(" ").map(Number);
            if (values.length !== 2 || values.some(isNaN) || values.some(v => v <= 0)) {
                console.log("❌ Invalid input! Enter exactly TWO positive numbers.");
                resolve(getMatrixDimensions(question)); // Retry input
            } else {
                resolve(values);
            }
        });
    });
}

// Function to read matrix elements
function getMatrix(rows, cols, name) {
    console.log(`Enter elements of ${name} matrix (space-separated, row by row):`);
    return new Promise(resolve => {
        let matrix = [];
        let rowCount = 0;

        const readRow = () => {
            rl.question("", input => {
                let row = input.trim().split(" ").map(Number);
                if (row.length !== cols || row.some(isNaN)) {
                    console.log(`❌ Please enter exactly ${cols} numbers for this row!`);
                    return readRow(); // Retry input
                }
                matrix.push(row);
                rowCount++;

                if (rowCount < rows) readRow();
                else resolve(matrix);
            });
        };

        readRow();
    });
}

// Main function to perform matrix multiplication
async function main() {
    let [m, n] = await getMatrixDimensions("Enter rows and columns of first matrix: ");
    let [n2, p] = await getMatrixDimensions("Enter rows and columns of second matrix: ");

    if (n !== n2) {
        console.log("❌ Matrix multiplication is not possible! Columns of first matrix must match rows of second.");
        rl.close();
        return;
    }

    let A = await getMatrix(m, n, "First");
    let B = await getMatrix(n, p, "Second");

    // Initialize result matrix C
    let C = Array.from({ length: m }, () => Array(p).fill(0));

    // Perform multiplication
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < p; j++) {
            for (let k = 0; k < n; k++) {
                C[i][j] += A[i][k] * B[k][j];
            }
        }
    }

    console.log("\n✅ Resultant Matrix:");
    C.forEach(row => console.log(row.join(" ")));

    rl.close();
}

// Run the program
main();

File: ARRAY/Multiplication/Multiplication.py
def get_matrix_dimensions(prompt):
    while True:
        try:
            values = input(prompt).split()
            if len(values) != 2:
                print("Please enter exactly TWO numbers separated by a space (e.g., '2 3')")
                continue  # Ask for input again
            m, n = map(int, values)
            if m <= 0 or n <= 0:
                print(" Dimensions must be positive integers! Try again.")
                continue
            return m, n
        except ValueError:
            print("Invalid input! Enter valid integers (e.g., '2 3').")

# Get matrix dimensions
m, n = get_matrix_dimensions("Enter rows and columns of first matrix: ")
p = int(input("Enter columns of second matrix: "))

# Ensure p is a valid integer
while p <= 0:
    print("Matrix dimensions must be positive integers! Try again.")
    p = int(input("Enter columns of second matrix: "))

print(f"Matrix 1: {m}x{n}, Matrix 2: {n}x{p}")

# Function to input a matrix
def input_matrix(rows, cols, name):
    print(f"Enter elements of {name} matrix ({rows}x{cols}):")
    matrix = []
    for i in range(rows):
        row = list(map(int, input().split()))
        while len(row) != cols:  # Ensure correct number of elements
            print(f"Please enter exactly {cols} numbers for row {i+1}!")
            row = list(map(int, input().split()))
        matrix.append(row)
    return matrix

# Read matrices
A = input_matrix(m, n, "First")
B = input_matrix(n, p, "Second")

# Perform matrix multiplication
C = [[0] * p for _ in range(m)]  # Initialize result matrix

for i in range(m):
    for j in range(p):
        for k in range(n):
            C[i][j] += A[i][k] * B[k][j]

# Display Resultant Matrix
print(" Resultant Matrix (Product):")
for row in C:
    print(" ".join(map(str, row)))
